src/som/vm/ExtensionLoader.java

```diff
@@ -0,0 +1,134 @@
+import com.oracle.truffle.api.source.Source;
+import com.oracle.truffle.api.source.SourceSection;
+
+import bd.primitives.Primitive;
+import bd.primitives.PrimitiveLoader;
+import bd.primitives.Specializer;
+import som.VM;
+import som.compiler.AccessModifier;
+import som.compiler.MethodBuilder;
+import som.interpreter.SomLanguage;
+import som.interpreter.nodes.ArgumentReadNode.LocalArgumentReadNode;
+import som.interpreter.nodes.ExpressionNode;
+import som.interpreter.nodes.dispatch.Dispatchable;
+import som.vmobjects.SInvokable;
+import som.vmobjects.SSymbol;
+
+
+public final class ExtensionLoader extends PrimitiveLoader<VM, ExpressionNode, SSymbol> {
+
+  private final EconomicMap<SSymbol, Dispatchable> primitives;
+
+  private final String         moduleName;
+  private final SomLanguage    lang;
+  private final URLClassLoader moduleJar;
+  private final String         extensionClassName;
+  private final Extension      extension;
+
+  public ExtensionLoader(final String moduleName, final SomLanguage lang) {
+    super(Symbols.PROVIDER, lang.getVM());
+    primitives = EconomicMap.create();
+    this.lang = lang;
+    this.moduleName = moduleName;
+
+    moduleJar = loadJar(moduleName);
+    extensionClassName = getExtensionClassName(moduleJar);
+    extension = loadExtensionClass();
+
+    initialize();
+  }
+
+  private URLClassLoader loadJar(final String moduleName) {
+    File f = new File(moduleName);
+    assert f.exists();
+
+    try {
+      return new URLClassLoader(new URL[] {f.toURI().toURL()},
+          getClass().getClassLoader());
+    } catch (MalformedURLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private String getExtensionClassName(final URLClassLoader moduleJar) {
+    Properties p = new Properties();
+    try {
+      p.load(moduleJar.getResourceAsStream("somns.extension"));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    return p.getProperty("class");
+  }
+
+  private Extension loadExtensionClass() {
+    Class<?> ext;
+    try {
+      ext = moduleJar.loadClass(extensionClassName);
+      return (Extension) ext.newInstance();
+    } catch (IllegalAccessException | InstantiationException | ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public EconomicMap<SSymbol, Dispatchable> getPrimitives() {
+    return primitives;
+  }
+
+  @Override
+  protected List<NodeFactory<? extends ExpressionNode>> getFactories() {
+    return extension.getFactories();
+  }
+
+  @Override
+  protected void registerPrimitive(final Primitive prim,
+      final Specializer<VM, ExpressionNode, SSymbol> specializer) {
+    String name = prim.primitive();
+
+    assert !("".equals(name));
+    SSymbol signature = Symbols.symbolFor(name);
+    assert !primitives.containsKey(signature) : "clash of vmMirrorPrimitive names";
+    primitives.put(signature,
+        constructPrimitive(signature, specializer, lang));
```

Comment:
Should we prefix primitives by their module names to avoid potential clashes?

Reply:
There are no clashes possible. They are entirely isolated from each other.

Reply:
If we are correct, there can be clash only between primitive signatures within a plugin, not between the primitive signature in 1 plugin with the signature in vmMirror or in another plugin.

So why the error msg "assert !primitives.containsKey(signature) : "clash of vmMirrorPrimitive names";"?

A different message could be better, for example ("clash of primitive named " + name  + " in " + moduleName).

This looks important when writing the module to know what is problematic (typically when copy pasting one primitive, modifying the copied version to new code and forgetting to change the primitive signature).